pkg/apis/clusterregistry/v1alpha1/types.go

```diff
@@ -167,3 +171,81 @@ type ClusterCondition struct {
 	// +optional
 	Message string `json:"message,omitempty" protobuf:"bytes,6,opt,name=message"`
 }
+
+// ClusterConnectionSpec defines the desired state of a ClusterConnection.
+type ClusterConnectionSpec struct {
+	// ClusterRef is a reference to a Cluster whose API endpoints should be
+	// connected to.
+	ClusterRef v1.LocalObjectReference `json:"clusterRef,omitempty"`
+
+	// Name of the secret in the namespace that contains kubeconfig to access
+	// the referenced cluster. The Secret have a kubeconfig in its data with key
+	// "kubeconfig".
+	//
+	// This can be left empty if the cluster allows insecure access.
+	// +optional
+	SecretRef *apiv1.LocalObjectReference `json:"secretRef,omitempty"`
+}
+
+// ClusterConnectionStatus contains information about the current status of a
+// ClusterConnection.
+type ClusterConnectionStatus struct {
+	// Conditions is an array of current conditions.
+	// +optional
+	Conditions []ClusterConnectionCondition `json:"conditions,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterConnection represents a connection to a Cluster using a Kubeconfig.
+//
+// +k8s:openapi-gen=true
+// +kubebuilder:resource:path=clusterconnections
+// +kubebuilder:subresource:status
+type ClusterConnection struct {
```

Comment:
Note: I do not like this name but didn't want to spend a lot of time naming before a group discussion